File: shared_data/si_unit.rb
# this class a SI unit https://en.wikipedia.org/wiki/International_System_of_Units
class SiUnit
    # defines one value.
    # second, meter, etc. are defined as powers
    # example: Joule is kg⋅m^2⋅s^−2 => kg = 1, m = 2, s = -2, A = 0, K = 0, mon = 0, cd = 0
    def initialize(s, m, kg, A, K, mol, cd)
        # powers as si units
        @second = s # example: 2 for s², -2 for 1/s²
        @metre = m
        @kilogram = kg
        @ampere = A
        @kelvin = K
        @mole = mol
        @candela = cd
    end
end

File: client/Research2DPrototype/Assets/Scripts/SI/SiUnit.cs
using System;

public class SiUnit
{
  public int Second { get; private set; }
  public int Metre { get; private set; }
  public int Kilogram { get; private set; }
  public int Ampere { get; private set; }
  public int Kelvin { get; private set; }
  public int Mole { get; private set; }
  public int Candela { get; private set; }

  public SiUnit()
  {
  }
  public SiUnit(int s, int m, int kg, int A, int K, int mol, int cd)
  {
    this.Second = s;
    this.Metre = m;
    this.Kilogram = kg;
    this.Ampere = A;
    this.Kelvin = K;
    this.Mole = mol;
    this.Candela = cd;
  }

  public static SiUnit operator *(SiUnit a, SiUnit b)
  {
    return new SiUnit(a.Second + b.Second,
      a.Metre + b.Metre,
      a.Kilogram + b.Kilogram,
      a.Ampere + b.Ampere,
      a.Kelvin + b.Kelvin,
      a.Mole + b.Mole,
      a.Candela + b.Candela);
  }

  public static SiUnit operator /(SiUnit a, SiUnit b)
  {
    return new SiUnit(a.Second - b.Second,
      a.Metre - b.Metre,
      a.Kilogram - b.Kilogram,
      a.Ampere - b.Ampere,
      a.Kelvin - b.Kelvin,
      a.Mole - b.Mole,
      a.Candela - b.Candela);
  }
}